source/transfer_list.rst

```diff
@@ -164,28 +198,20 @@ The `tag_id` ranges are described in :numref:`tab_tag_id_ranges`.
    * - tag ID range
      - Description
 
-   * - 0x0 -- 0xf_ffff
-     - Standard tag id range. Any tag id in this range must first be allocated in this specification before being used. The allocation of the tag id requires the entry layout to be defined as well.
+   * - 0x0 -- 0x7f_ffff
+     - Standardized range. Any tag ID in this range must first be allocated in this specification before being used. The allocation of the tag ID requires the entry layout to be defined as well.
 
+   * - 0x80_0000 -- 0xff_efff
+     - Reserved. (Can later be used to extend standardized range if necessary.)
 
-   * - 0x10_0000 -- 0x10_ffff
-     - Non-standard range. A platform firmware integrator can create entries in this range. Different platforms are allowed to have tag ids in this range with distinct data formats. Entries in this range are not standardized.
-
-   * - 0x11_0000 -- 0xffff_ffff
-     - Reserved
+   * - 0xff_f000 -- 0xff_ffff
+     - Non-standardized range. Tag IDs in this range can be used without allocation in this specification. Using this range for anything other than local experimentation or closed-source components that are entirely under the control of a single platform firmware integrator is strongly discouraged.
 
 .. _sec_std_entries:
 
 Standard transfer entries
 -------------------------
 
-The TEs have a `tag_id` in the {0, ..., 0xf_ffff} set. Both
-the tag_id of a standard entry as well as the entry layout
-must be defined in this specification before being used.
-New entries are expected to have a simple layout. Complex
-data should be represented in a self-describing data
```

Comment:
Please keep this last sentence

Reply:
Well, I'm not really sure how to fit that single sentence into the rewrite. I also feel it is ill-defined... what is "complex data"? This is what I tried to describe more clearly in the second paragraph of my new text ("Tags are expected to have a simple layout (representable by a C structure) and each tag should only represent data for a single logical concept. ..."). Doesn't that already disallow "complex data"? If not please explain what kind of scenario you think shouldn't be allowed that you think isn't covered by my text and I'll try to figure out how to fit that in.

Reply:
You can just leave it  as is, if you like.

But if you want to expand it...

Complex data is anything more than a simple C struct. So for example, ACPI and SMBIOS would be complex. Multiple C structs would be complex. Perhaps we should set a field limit as well, so people don't use it for FSP parameters.

How about:

Complex data, such as multiple C structs or structs with more than 5-10 fields, should be represented in a self-describing data structure, such as the FDT [DT]_

Reply:
I don't think it makes sense to come up with a hard field limit, that's why I tried to explain this as "should represent single logical concept" instead (and the actual determination can then be made by reviewers on a case by case basis). For example, the framebuffer information structure coreboot passes to its payloads contains 14 distinct fields, but I don't think that means it would be too "complex" for this data structure... it's just defining the bit offset and size of each color channel in the packed pixel representation in an extensible way, and the cleanest way to do that is with a separate mask and offset field for each channel.

Not sure what exactly you mean by "multiple C structs". I think the case of having a simple array of a sub struct should be allowed, because it is very useful and comes up often, e.g. something like:
```
struct timestamp_table_tag {
  uint32_t num_entries;
  struct timestamp_entry {
    uint32_t entry_id;
    uint32_t entry_value;
  }[];
}
```
So I think that's roughly the limit of what should be allowed and I think that's covered by my text (it is still "representable by a C structure", i.e. my definition of `struct timestamp_table_tag` above is valid C code). A construct where, say, multiple structures are placed back-to-back and some of them are optional based on values set in the first structure would not be allowed because that's not "representable by a C structure" anymore. (Of course, wrapping a complex legacy structure in a tag as an opaque byte array is still allowed because that's what the HOB, FDT and ACPI tags are already doing, but that should not be done for new structures created exclusively for the transfer list.)

Reply:
I think it would faster to come to agreement if you respond to what people say, rather than inferring things that were not said. I dd not suggest a 'hard' field limit. 

A (soft) field limit is useful as it provides guidance as to when you are off the reservation. The example you show makes sense I suppose. What happens when you want to add more entries to the timestamp log? I suppose you create another struct with the same tag later on in the TL.

It seems we mostly agree with the intent. My goal is to actually express this in text, so please suggest an alternative to what I wrote.

Reply:
1. I'd like to see guidance on the number of members in the structs, e.g. no more than 5-10 fields. It will save us arguing about it in PR reviews
2. The bit about complex structures using FDT needs to stay as it is important to explain the fallback

Reply:
Well, I really disagree with calling out any specific number of fields, because as I have tried to explain and illustrate with examples earlier, this metric is really meaningless when some things that belong together just require a bunch of fields to describe while on the other hand I can easily write a TE that only contains 3 fields but they all concern completely independent things that should be described in an independent TE each. We need a more abstract rule to describe this restriction.

We have discussed the issue with the FDT comment to death already. This is supposed to be a universal handoff specification with guidelines about how things should be represented but no restrictions on how projects are allowed to use it or what they are allowed to represent. This is not just supposed to be a pre-RAM substitute for FDT. It should not tell users to use any specific other data structure instead, it should explain how to properly represent any kind of information in this structure.

Earlier you commented "You can just leave it as is, if you like.", but it sounds like you are retracting that now? @jmarinho , @danh-arm , do you have opinions on this topic that might help us resolve this impasse?

Reply:
I've been avoiding commenting here because it feels to me like we're getting into bike shedding territory. I don't feel strongly about this; both Jose and I have approved this version so we could merge it now, although it would be nice to have general consensus. For what it's worth, my opinion is that I agree with Julius that guidance on the max number of fields seems unnecessarily restrictive. At the same time, I think the existing words to use something like FDT for complex types is OK. I think I suggested in a previous review that maybe this wording could be softened to achieve consensus, e.g. "For complex data, consider the use of a self-describing data structure, for example FDT [or HOB or ACPI]".

Reply:
I really feel that this spec should contain no suggestions of the form "for XXX just use <something else> instead". I think a universal firmware handoff specification should be able to serve all firmware handoff needs. Otherwise, what's the point in trying to make something new that's focused on interoperability if half the stuff you might want to interoperate on is not covered by it and has to be represented by something else that requires some other parser again? We should not be precluding anyone from mixing this with other data structures if they want to but I don't think we should be suggesting it as the standard solution either. Those projects that want to live purely based on this specification should be able to do so without needing a fallback for specific kinds of data.

I fully understand the "tag layouts should not be as complex as an ACPI table" argument, and I've been trying to write guidelines in this PR that should rule that out (I just don't think specifying it in number of fields is a good approach). But I think the correct solution to the "complex data" problem is to break it down into multiple pieces of simple data that are passed as multiple TEs. If the assertion here is that there exist some kinds of data which firmware might want to pass between stages that could never possibly be broken down into a form that would be appropriate for the TL and will always require wrapping in some other data structure, then I fundamentally disagree with that. This spec should be able to stand on its own and offer at least one correct option to solve every firmware handoff use case by itself.

Reply:
> But I think the correct solution to the "complex data" problem is to break it down into multiple pieces of simple data...

That is one valid solution but I don't think that should be the goal of this specification. In systems that have a normal world bootloader operating on data in a particular format (FDT, HOB, etc..) and has earlier firmware operating on the same data, it _may_ be a better design choice to keep the data in that format throughout, rather than doing conversions.

This isn't worth arguing about, though. Let's leave the sentence out - it isn't of material importance.

Reply:
Right, I in no way want to preclude systems where that makes sense to use it like this (and I think the fact that the FDT TE is one of the first defined in this spec makes it clear that wrapping those formats is a valid option). But if there's a sentence that just says "complex data structures need to be represented in an FDT", that's not an option, that's a requirement. I don't want this to later be used in reviews to force an FDT dependency onto projects that don't want to have one.

Would it work as a compromise if I add a new sentence like this: "Data for multiple distinct concepts should be split across different tags, [...]. _Alternatively, complex data may be represented in a different kind of well-established handoff data structure (e.g. FDT, ACPI table) that is inserted into the TL as a single TE._"?

Reply:
Yes, your new sentence works for me!

Reply:
Updated (replaced "ACPI table" with "HOB" in the example).